fix: fix bug with customheadersoption

eliekozah opened the pull request:
This is a Draft PR for PR issue b/402481452

Please do not merge.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/googleapis/google-cloud-cpp/15053)
<!-- Reviewable:end -->

ddelgrosso1 commented:
A few general comments, it looks like this is breaking at least one test in the `asan-pr` build. Also, looking at CLang tidy there are some other problems with copy construction there.

That said, as we discussed I'm not sure adding a new function is the way to best fix this. Perhaps this is a good topic to raise in the weekly C++ meeting tomorrow but I would think the existing functions should behave correctly.

ddelgrosso1 commented on google/cloud/storage/internal/rest/stub.cc:
```diff
@@ -106,6 +106,36 @@ StatusOr<ReturnType> CreateFromJson(
+    for (auto const& h : options.get<CustomHeadersOption>()) {
+      builder.AddHeader(h.first, h.second);
+    }
+    return {};
```
I'm surprised this doesn't cause an error as the function is declared returning a `Status`. I would check what this actually does when it returns.

cuiy0006 commented on google/cloud/storage/internal/rest/stub.cc:
```diff
@@ -106,6 +106,36 @@ StatusOr<ReturnType> CreateFromJson(
+
+void AddOptionsToRequestBuilder(Options const& options,
+  RestRequestBuilder& builder) {
+    Status cust = AddCustomHeaders(options, builder);
```
If we don't care about status, then remove it. Otherwise, we will want to add a condition or return it to its caller.

cuiy0006 commented on google/cloud/storage/internal/rest/stub.cc:
```diff
@@ -167,7 +197,7 @@ StatusOr<ListBucketsResponse> RestStub::ListBuckets(
   auto auth = AddAuthorizationHeader(options, builder);
   if (!auth.ok()) return auth;
-  request.AddOptionsToHttpRequest(builder);
+  AddOptionsToRequestBuilder(options, builder);
```
I guess `request.AddOptionsToHttpRequest(builder)` is to add info from request to the builder, after change, is the info from request missed? [Example](https://github.com/googleapis/google-cloud-cpp/blob/f080acd0c88b15f1d505b391c60aa69ac29849ff/google/cloud/storage/internal/rest/request_builder_test.cc#L42-L58)

eliekozah commented:
I deleted the previous "quick fix" and added a new class CustomHeaders in GCS.
You can now add **multiple** custom headers to your request.

example:
```
    auto response = client.ListObjects(
        "my_bucket",
        gcs::CustomHeaders({{"name1", "value1"}, {"name2", "value2"}})
    );
```

scotthart reviewed (changes requested):
Can we add some unit tests that verify the desired behavior?

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/15053#-:-OMqxotM81EdMSPB4XwV:b-soy2ht)* status: 0 of 8 files reviewed, 5 unresolved discussions (waiting on @cuiy0006, @ddelgrosso1, and @eliekozah)

___
*[`google/cloud/storage/internal/rest/stub.cc` line 121 at r4](https://reviewable.io/reviews/googleapis/google-cloud-cpp/15053#-OMqx9Gu3W_H7silc-F2:-OMqx9Gu3W_H7silc-F3:bybtdfb) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/c3a91666befe906bc4b1ee4a092df3cb59d0fe91/google/cloud/storage/internal/rest/stub.cc#L121)):*
> ```cpp
> }
> 
> Status AddCustomHeaders(Options const& options, RestRequestBuilder& builder) {
> ```

It doesn't appear that anything called in this function can fail, so let's change the return type to `void`.
___
*[`google/cloud/storage/internal/rest/stub.cc` line 129 at r4](https://reviewable.io/reviews/googleapis/google-cloud-cpp/15053#-OMqxTA10TKwYs8pQ5-w:-OMqxTA10TKwYs8pQ5-x:bt6uwme) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/c3a91666befe906bc4b1ee4a092df3cb59d0fe91/google/cloud/storage/internal/rest/stub.cc#L129)):*
> ```cpp
> }
> 
> Status AddHeaders(Options const& options, RestRequestBuilder& builder) {
> ```

It appears `AddAuthortizationHeader` can fail, so this function needs to continue to return `Status`. Additionally, calls to this function need to propagate up any non-OK `Status` that are returned.


<!-- Sent from Reviewable.io -->

ddelgrosso1 commented:
@eliekozah can you convert this to a draft pr if you are still working on it. Will help avoid people reviewing it.

eliekozah reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/15053)* status: 0 of 10 files reviewed, 5 unresolved discussions (waiting on @cuiy0006, @ddelgrosso1, and @scotthart)

___
*[`google/cloud/storage/internal/rest/stub.cc` line 121 at r4](https://reviewable.io/reviews/googleapis/google-cloud-cpp/15053#-OMqx9Gu3W_H7silc-F2:-ONFdv2S4NWIk8uqY1Xw:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/c3a91666befe906bc4b1ee4a092df3cb59d0fe91/google/cloud/storage/internal/rest/stub.cc#L121)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

It doesn't appear that anything called in this function can fail, so let's change the return type to `void`.
</blockquote></details>

Done.
___
*[`google/cloud/storage/internal/rest/stub.cc` line 129 at r4](https://reviewable.io/reviews/googleapis/google-cloud-cpp/15053#-OMqxTA10TKwYs8pQ5-w:-ONFdyESEGxF7bZjb8lw:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/c3a91666befe906bc4b1ee4a092df3cb59d0fe91/google/cloud/storage/internal/rest/stub.cc#L129)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

It appears `AddAuthortizationHeader` can fail, so this function needs to continue to return `Status`. Additionally, calls to this function need to propagate up any non-OK `Status` that are returned.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ddelgrosso1 commented on google/cloud/storage/internal/rest/stub.cc:
```diff
@@ -118,7 +120,19 @@ Status AddAuthorizationHeader(Options const& options,
   return {};
 }
 
-}  // namespace
```
It looks like you moved AddAuthorizationHeader out of the namespace. Was there a reason to move it?

eliekozah replied:
I wanted to make it accessible for a stub_test.cc unit test.

eliekozah replied:
I'll remove one of the 2 unit test I added that calls the function directly. I don't think there's a need for it.

ddelgrosso1 commented on google/cloud/storage/internal/rest/stub.cc:
```diff
@@ -165,8 +179,7 @@ StatusOr<ListBucketsResponse> RestStub::ListBuckets(
       absl::StrCat("storage/", options.get<TargetApiVersionOption>(), "/b"));
-  auto auth = AddAuthorizationHeader(options, builder);
-  if (!auth.ok()) return auth;
+  AddHeaders(options, builder);
```
You probably should be checking the status returned from AddHeaders just like it was doing for AddAuthorizationHeader.

ddelgrosso1 commented on google/cloud/storage/internal/rest/stub.h:
```diff
@@ -29,6 +30,8 @@ namespace storage {
 GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
 namespace internal {
 
+void AddCustomHeaders(Options const& options, RestRequestBuilder& builder);
```
No need for this in the .h, can just define it within the C++ similar to AddAuthorizationHeader

eliekozah replied:
I wanted to make it accessible for a stub_test.cc unit test. I just removed the declaration as there's no more need for one of the unit tests.

ddelgrosso1 commented on google/cloud/storage/internal/rest/stub.cc:
```diff
@@ -165,7 +179,7 @@ StatusOr<ListBucketsResponse> RestStub::ListBuckets(
   RestRequestBuilder builder(
       absl::StrCat("storage/", options.get<TargetApiVersionOption>(), "/b"));
-  auto auth = AddAuthorizationHeader(options, builder);
+  auto auth = AddHeaders(options, builder);
```
I would suggest changing the name of this variable as it is no longer just the auth header.

ddelgrosso1 commented on google/cloud/storage/internal/rest/stub.cc:
```diff
@@ -118,6 +118,20 @@ Status AddAuthorizationHeader(Options const& options,
   return {};
 }
 
+void AddCustomHeaders(Options const& options, RestRequestBuilder& builder) {
```
I would probably have this function return a `Status` just like AddAuthorization header, then in `AddHeaders` do a similar status check.

bajajneha27 commented on google/cloud/storage/internal/rest/stub.cc:
```diff
@@ -118,6 +118,20 @@ Status AddAuthorizationHeader(Options const& options,
+}
+
+Status AddHeaders(Options const& options, RestRequestBuilder& builder) {
+  auto ah = AddAuthorizationHeader(options, builder);
```
Just curious, do we need to necessarily do this before adding any header?

eliekozah replied:
So far I think the answer is yes. `AddAuthorizationHeader` is called in every requests in stub.cc. Other headers (such Custom Headers) are optional.

bajajneha27 commented on google/cloud/storage/internal/rest/stub_test.cc:
```diff
@@ -144,6 +143,31 @@ Matcher<std::vector<absl::Span<char const>> const&> ExpectedPayload() {
+                    Contains(Pair("custom-header-2",
+                                  std::vector<std::string>{"value2"})));
+
+        return PermanentError();
```
I think we should add tests for success scenario as well.

eliekozah replied:
@bajajneha27 In this test I deliberately use a dummy error (PermanentError) to simulate a network failure so that I focus only on verifying that the header insertion logic behaves as expected. I bypassed the need to construct a full valid HTTP response by forcing an error.

Given that this test is focused only on checking the header propagation, do you think we should add an additional test that simulates a successful response as well?

bajajneha27 replied:
No, this sounds good. My bad.

ddelgrosso1 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/15053#-:-ONvLzNh0lgRvEp0AELn:bnfp4nl)* status: 0 of 10 files reviewed, 13 unresolved discussions (waiting on @bajajneha27, @cuiy0006, @eliekozah, and @scotthart)



<!-- Sent from Reviewable.io -->

scotthart reviewed (approved):
> Reviewed 1 of 3 files at r6.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/15053)* status: 1 of 10 files reviewed, 13 unresolved discussions (waiting on @bajajneha27, @cuiy0006, and @eliekozah)



<!-- Sent from Reviewable.io -->
